Scripts/Generator/RoomGenerator.cs

```diff
@@ -128,4 +169,74 @@ public PhysicsBody GenerateWall(Vector3 wallBase, Vector3 size, Vector3 normal)
         wall.Rotation = new Vector3(0, 1, 0) * normal.AngleTo(Vector3.Forward);
         return wall;
     }
+
+    public void GenerateFloor()
+    {
+        if (floor == null)
+            return;
+        Debug.Assert(tileOffset >= 0 && tileOffset <= 1);
+        Debug.Assert(tileSize.Length() > 0);
+        foreach (Node tile in floor.GetChildren())
+        {
+            if (tile != null)
+                tile.QueueFree();
+        }
+        Vector2 roomSpaceSize = roomSize - 2 * wallThickness * Vector2.One;
+        Vector3 startingCorner = origin - roomSpaceSize.x / 2 * left - roomSpaceSize.y / 2 * front;
+        for (int x = 0; x < roomSpaceSize.x / tileSize.x; x++)
+        {
+            float rowTileOffset = (tileOffset * x) % 1;
+            int y = 0;
+            if (rowTileOffset > 0)
+            {
+                y = -1;
+            }
+            while (y < roomSpaceSize.y / tileSize.y)
+            {
+                Vector2 currentTileOffset = new Vector2(x * tileSize.x, (y + rowTileOffset) * tileSize.y);
+                Vector2 currentTileSize = new Vector2(tileSize);
+                if (currentTileOffset.y < 0)
+                {
+                    currentTileSize.y += currentTileOffset.y;
+                    currentTileOffset.y = 0f;
+                }
+                if (currentTileOffset.y + tileSize.y > roomSize.y - wallThickness)
+                {
+                    float extraHeight = currentTileOffset.y + tileSize.y - (roomSize.y - wallThickness);
+                    currentTileSize.y -= extraHeight;
+                }
+                if (currentTileOffset.x + tileSize.x > roomSize.x - wallThickness)
+                {
+                    float extraWidth = currentTileOffset.x + tileSize.x - (roomSize.x - wallThickness);
+                    currentTileSize.x -= extraWidth;
+                }
+
+                Mesh tileMesh = new PlaneMesh()
+                {
+                    Size = currentTileSize
+                };
+                MeshInstance meshInstance = new MeshInstance()
+                {
+                    Mesh = tileMesh
+                };
+
+                Vector3 currentTileCorner = startingCorner + currentTileOffset.x * left + currentTileOffset.y * front + 0.01f * Vector3.Up;
+
+                // Place tile in the center instead of the corner.
+                Vector3 translation = currentTileCorner + (currentTileSize.x / 2) * left + (currentTileSize.y / 2) * front;
+                meshInstance.Translation = translation;
+
+                SpatialMaterial floorMaterial = new SpatialMaterial()
+                {
+                    AlbedoColor = Colors.White.LinearInterpolate(Colors.Black, (currentTileOffset.x / roomSize.x + currentTileOffset.y / roomSize.y) * 0.5f)
+                };
+                meshInstance.SetSurfaceMaterial(0, floorMaterial);
+
+                floor.AddChild(meshInstance);
+
+                y++;
```

Comment:
use for loop instead?

Reply:
I did use, but it will be like `for (; y< blabla; y++)`, if you don't mind.

Reply:
easier to see what happens to y right, otherwise y at the bottom of the whole block

Reply:
true that